src/chocolatey/infrastructure.app/runners/ListRunner.cs

```diff
@@ -0,0 +1,113 @@
+﻿// Copyright © 2015 - Present RealDimensions Software, LLC
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// 
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace chocolatey.infrastructure.app.runners
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using adapters;
+    using attributes;
+    using configuration;
+    using logging;
+    using infrastructure.commands;
+    using SimpleInjector;
+    using Console = System.Console;
+    using Environment = System.Environment;
+    public sealed class ListRunner<T>
+    {
+        public IEnumerable<T> List(ChocolateyConfiguration config, Container container, bool isConsole, Action<ICommand> parseArgs, bool logResults = true)
+        {
+            var commands = container.GetAllInstances<ICommand>();
+            var command = commands.Where((c) =>
+            {
+                var attributes = c.GetType().GetCustomAttributes(typeof(CommandForAttribute), false);
+                return attributes.Cast<CommandForAttribute>().Any(attribute => attribute.CommandName.is_equal_to(config.CommandName));
+            }).FirstOrDefault();
+
+            if (command == null)
+            {
+                //todo add a search among other location/extensions for the command
+                if (!string.IsNullOrWhiteSpace(config.CommandName))
+                {
+                    throw new Exception("Could not find a command registered that meets '{0}'".format_with(config.CommandName));
+                }
+
+                if (isConsole) Environment.ExitCode = 1;
+            }
+            else
+            {
+                var listCommand = command as IListCommand<T>;
+                if (listCommand == null)
+                {
+                    throw new Exception("The implementation of '{0}' does not support listing '{1}'".format_with(config.CommandName, typeof(T).Name));
+                }
+
+                if (parseArgs != null)
+                {
+                    parseArgs.Invoke(listCommand);
+                }
+
+                this.Log().Debug(() => "Configuration: {0}".format_with(config.ToString()));
+
+                if (isConsole && config.HelpRequested)
+                {
+#if DEBUG
+                    Console.WriteLine("Press enter to continue...");
+                    Console.ReadKey();
+#endif
```

Comment:
All except the last few lines of this are copy/paste. I thought about just adding stuff to the existing GenericRunner, but at the time when I did this I couldn't imagine a way to extend the configuration without creating dozens of combinations which would be incorrect...